### PR TITLE
Add a pipeline export command

### PR DESCRIPTION
**Pull request by chmouel**
Will export the pipeline definition to a yaml file, removing the auto generated
fields.

If you do a `tkn pipeline export pipeline` it will convert a pipeline that look like this : 

https://paste.gg/p/anonymous/e105629cffe547238f125f0862aa23be

and export it like that : 

https://paste.gg/p/anonymous/7515299fddbe44b3b54aaf12e6a7bba1

Partially address bug #1220

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Run the code checkers with `make check`
- [ ] Regenerate the manpages, docs and go formatting with `make generated`
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/cli/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

### CODE REVIEW

**Comment by chmouel** [comment]
Probably should discuss this next,

I am torn between : 

`tkn export pipeline`

or like this is implemented here : 

`tkn pipeline export`

cc @nikhil-thomas @siamaksade  @tektoncd/cli-maintainers

**Comment by chmouel** [comment]
OOps just saw there is @vinamra28 https://github.com/tektoncd/cli/pull/1359#partial-pull-merging which is not the same but pretty similiar, either way I can update this PR post 1359 or close this if 1359 addresse this use case (exporting a task as yaml)

File: pkg/cmd/pipeline/export.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+func exportCommand(p cli.Params) *cobra.Command {
+	f := cliopts.NewPrintFlags("export")
+
+	eg := `List all PipelineRuns of Pipeline 'foo':
+
+    tkn pipelinerun list foo -n bar
+
+List all PipelineRuns in a namespace 'foo':
+
+    tkn pr list -n foo
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:30]: can you please change this for export command ? :stuck_out_tongue:

File: pkg/cmd/pipeline/export.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/tektoncd/cli/pkg/options"
+	"github.com/tektoncd/cli/pkg/pipeline"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cliopts "k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+type ExportOptions struct {
+	Params       cli.Params
+	PipelineName string
+}
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:19]: I see that this struct is not used anywhere. Can you please remove it?

**Review by vinamra28 — COMMENTED** [review]
can we have a unit test for this command as well in order to maintain the consistency?

**Comment by chmouel** [comment]
Yup will work on this

File: pkg/cmd/pipeline/export.go (hunk 2 above)
Comment (chmouel) [reply pkg/cmd/pipeline/export.go:19]: the plan is to move this to common namespace and use this for Task/ClusterTask export when I or someone else get a bit more time to implement it,

File: pkg/cmd/pipeline/export.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+func exportCommand(p cli.Params) *cobra.Command {
+	f := cliopts.NewPrintFlags("export")
+
+	eg := `List all PipelineRuns of Pipeline 'foo':
+
+    tkn pipelinerun export foo -n bar
+
+List all PipelineRuns in a namespace 'foo':
+
+    tkn pr export -n foo
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:30]: nit: can you please change this for export command?

File: pkg/cmd/pipeline/export.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			}
+			fmt.Println(exported)
+			return nil
+		},
+	}
+	f.AddFlags(c)
+	return c
+}
+
+func exportPipeline(out io.Writer, p cli.Params, pname string) (string, error) {
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:71]: are we using the param value `out io.Writer` anywhere?

File: pkg/cmd/pipeline/export.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+				}
+			} else {
+				opts.PipelineName = args[0]
+			}
+
+			exported, err := exportPipeline(cmd.OutOrStdout(), p, opts.PipelineName)
+			if err != nil {
+				return err
+			}
+			fmt.Println(exported)
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:63]: don't we need to write the output to `io.Writer` or is it fine with this as well? :sweat_smile:

**Comment by chmouel** [comment]
tks for the review @vinamra28,  updated

**Comment by vinamra28** [comment]
@chmouel I guess you need to run `make generated` once

File: pkg/cmd/pipeline/export.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+	pipeline, err := pipeline.Get(cs, pname, metav1.GetOptions{}, p.Namespace())
+	if err != nil {
+		return err
+	}
+	exported, err := export.PipelineToYaml(pipeline)
+	if err != nil {
+		return err
+	}
+	out.Write([]byte(exported))
```
Comment (vinamra28) [inline pkg/cmd/pipeline/export.go:85]: linter is not happy here :sweat_smile: how about
```go
fmt.Fprint(out, exported)
```

Comment (vinamra28) [reply pkg/cmd/pipeline/export.go:85]: or maybe you need to handle the return values of function `func (io.Writer).Write(p []byte) (n int, err error)`

**Comment by chmouel** [comment]
Thanks @vinamra28  there was a lot of warnings when running locally with newer golangci-lint (cf: https://github.com/tektoncd/cli/issues/1481) and didn't see this one

**Review by vinamra28 — APPROVED** [review]
/lgtm
/meow

File: pkg/export/export.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+// pipelineExportRemoveExtraFields some of the fields are not getting skipped, when set as nil,
+// like creationTimeStamp since it's a time.Time that is not a pointer yaml would display a creationTimeStamp: null
+// let's do some regexp magic here, which should hopefully safe.
+func pipelineExportRemoveExtraFields(input string) string {
+	m1 := regexp.MustCompile(`\n  creationTimestamp: null\n`)
```
Comment (pradeepitm12) [inline pkg/export/export.go:15]: `\n  creationTimestamp: null\n`
Do we need two spaces in the regex

Comment (chmouel) [reply pkg/export/export.go:15]: Yep that's how it matches ,what do you suggest we do ? Having a \s* ?

Comment (pradeepitm12) [reply pkg/export/export.go:15]: My editor showed an error for this :smile_cat: just wanted to check.
/lgtm

**Review by pradeepitm12 — APPROVED** [review]

**Pull request merged**